[processor/schematransformerprocessor] Initial PR (#8371)
open-telemetry/opentelemetry-collector-contrib: processor/schemaprocessor/README.md

# Schema Transformer Processor

🚧 _Currently under development, subject to change_ 🚧

Supported Pipelines: traces, metrics, logs

The _Schema Processor_ is used to convert existing telemetry data or signals to a version of the semantic convention defined as part of the configuration. 

> The latest schema translation file will be bundled as part of the collector build to avoiding having to fetch it at runtime. 

Reviewer: The "latest" changes over time, specification publishes new versions and the "latest" schema release in Otel may happen after the particular bundle of the Collector is released. What happens in that case?
I think the Collector should be able to fetch the schema file on demand. Bundled "latest schema available at the release time may be an optimization to speed up Collector startup (but I am not totally sure it is necessary).

Author: I think it is reasonable to be not in complete lockstep, having the context on your idea of how you expect the processor to work, missing a schema translation file could be resolved at runtime.

However, this idea is a rather round about way and there is currently nothing set up to do the automation.

I should be clear that I read this idea from the spec for it and potentially considering moving this automation til after the ability to transform these signals is complete.

# Schema Transformer Processor

🚧 _Currently under development, subject to change_ 🚧

Supported Pipelines: traces, metrics, logs

The _Schema Processor_ is used to convert existing telemetry data or signals to a version of the semantic convention defined as part of the configuration.

Specification Reference: [here](https://opentelemetry.io/docs/reference/specification/schemas/overview/)

## Caching Schema Translation Files

In order to improve effeicency of the processor, the `precache` option allows the processor to start downloading and preparing
the translations needed for signals that match the schema url.

## Targets Schemas

Targets define a list of schema urls with a schema identifer that will be used to translate any schema url that matches the target url to that version.
In the event that the processor matches a signal to a target, the processor will translate the signal from published one to the defined identifier;

> using the configuration bellow, a signal published with `SchemaURL:https://opentelemetry.io/schames/1.8.2` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.2` so that all existing visualisations will remain unchanged.

Reviewer: ```suggestion
using the configuration bellow, a signal published with `SchemaURL:https://opentelemetry.io/schames/1.8.2` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.2`.
```
I think mentioning the reason why you would want to do a translation is nice to specify, but perhaps a better place for it would be the top of this document, after line 7.

Author: That makes sense to me!

> Furthemore, it is also possible for organisations and vendors to publish their own semantic convention and be used by this processor, be sure to follow [schema overview](https://opentelemetry.io/docs/reference/specification/schemas/overview/) for all the details.

Reviewer: I would also move this line somewhere closer to the top of the file.

# Example

```yaml
processors:
    schema_transformer:
        precache:
        - https://opentelemetry.io/schemas/1.9.0
        targets:

>         - https://opentelemetry.io/schemas/1.6.2

Reviewer: Nit: let's use URLs that actually exist:
```suggestion
        - https://opentelemetry.io/schemas/1.6.1
```

Author: Question for you, is there an easier way for a developer / curious person to see the list of published schemas?

Reviewer: Yes, all Otel schema file originals are here: https://github.com/open-telemetry/opentelemetry-specification/tree/main/schemas

# Schema Transformer Processor

🚧 _Currently under development, subject to change_ 🚧

Supported Pipelines: traces, metrics, logs

The _Schema Processor_ is used to convert existing telemetry data or signals to a version of the semantic convention defined as part of the configuration.

> The processor works by using a set of target schema URLs that are used to match incoming signal.

Reviewer 2: ```suggestion
The processor works by using a set of target schema URLs that are used to match the incoming signals.
```

> The processors works be using a set of target schema URLs that are used to match incoming signals. 

Reviewer: ```suggestion
The processor works by using a set of target schema URLs that are used to match incoming signals. 
```

> On a match, the processor will fetch the schema translation file (if not cached) set by the incoming signal and apply transformations

Reviewer 2: ```suggestion
On a match, the processor will fetch the schema translation file (if not cached) set by the incoming signal and apply the transformations
```

required to export as the target semantic convention version.

> Furthemore, it is also possible for organisations and vendors to publish their own semantic convention and be used by this processor, 

Reviewer 2: ```suggestion
Furthermore, it is also possible for organisations and vendors to publish their own semantic conventions and be used by this processor, 
```

be sure to follow [schema overview](https://opentelemetry.io/docs/reference/specification/schemas/overview/) for all the details.

## Caching Schema Translation Files

> In order to improve effeicency of the processor, the `precache` option allows the processor to start downloading and preparing

Reviewer 2: Is it too late to rename `precache` to `prefetch`? I think it's more widely used for similar features elsewhere.

Author: Makes sense to me :D

> In order to improve effeicency of the processor, the `precache` option allows the processor to start downloading and preparing

Author: Not sure if that was a mistake, but `misspell` is reporting it as a typo.

> the translations needed for signals that match the schema url.

Reviewer 2: ```suggestion
the translations needed for signals that match the schema URL.
```

## Targets Schemas

> Targets define a list of schema urls with a schema identifer that will be used to translate any schema url that matches the target url to that version.

Reviewer 2: ```suggestion
Targets define a list of schema URLs with a schema identifier that will be used to translate any schema URL that matches the target URL to that version.
```

> In the event that the processor matches a signal to a target, the processor will translate the signal from published one to the defined identifier;

Reviewer 2: ```suggestion
In the event that the processor matches a signal to a target, the processor will translate the signal from the published one to the defined identifier;
```

> using the configuration bellow, a signal published with `SchemaURL:https://opentelemetry.io/schemas/1.8.0` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.1`.

Reviewer 2: ```suggestion
using the configuration below, a signal published with `SchemaURL:https://opentelemetry.io/schemas/1.8.0` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.1`.
```

> using the configuration below, a signal published with `SchemaURL:https://opentelemetry.io/schemas/1.8.0` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.1`.

Reviewer 3: Wouldn't the example below translate signals published with *any* schema to 1.6.1? I'm not seeing where 1.8.0 is specified anywhere.

> using the configuration below, a signal published with `SchemaURL:https://opentelemetry.io/schemas/1.8.0` will be translated by the collector to `SchemaURL:https//opentelemetry.io/schemas/1.6.1`.

Author: I should have added "for example"
